### DISCUSSION TITLE
Using Docker for Suture Challenge + ROS interface

### DISCUSSION THREAD

**Discussion by sunshineinsandiego** [surgical-robotics-ai/surgical_robotics_challenge#75] [Q&A]
Hi - I am trying to setup my local machine (Ubuntu 20.04) with ambf and a docker container so I can interface with the suture challenge simulator using ROS scripts that I have written. 

I followed the step 1 instructions [here](https://github.com/collaborative-robotics/surgical_robotics_challenge) and set up ambf successfully on my local machine. 

I then followed step 2 [here](https://github.com/collaborative-robotics/surgical_robotics_challenge) and [here](https://github.com/collaborative-robotics/docker_surgical_robotics_challenge) and set up a docker image successfully on my local machine as well. 

I'm not entirely sure what to do next. 

Step 3 states "The simulation is spawned in AMBF with the launch file and AMBF Description Format (ADF) files from this repo" but if I'm using a docker container not sure how to proceed so that (1) I can visualize the simulator and (2) I can control the simulator via ROS scripts or input devices like the DVRK PSMs.

Thanks for your help!

**Comment by adnanmunawar** [comment]
Hello,

Thanks for the detailed description of your machine's state.

The docker container was provided for making it easy for people to submit entries for the surgical robotics challenge. In that approach, one would have cloned this repo as well as the AMBF on the local machine. They would then launch the simulation using the assets on the local machine. The docker container would really only contain the control scripts and the contents from the [scripts](https://github.com/collaborative-robotics/surgical_robotics_challenge/tree/master/scripts) folder from this repo. This would have allowed one to control the simulation via ROS on the local host.

You can ignore the docker step and directly execute your control scripts on the local machine and launch the tele-operation scripts to control the PSMs via the input devices.

**Comment by sunshineinsandiego** [comment]
Ah completely understand, thank you!

I did as directed and attempted to launch the tool challenge sim. I don't see any errors, but I'm also not seeing any PSM end effectors in the simulation.  This is what I see in the sim:

![ambf1](https://user-images.githubusercontent.com/56661595/236919890-8c27af69-7cda-4bb0-b406-0117b999bec1.png)
![ambf2](https://user-images.githubusercontent.com/56661595/236919894-2183b9a1-3e83-4b64-806b-6039935824f0.png)

**Comment by sunshineinsandiego** [comment]
Ah completely understand, thank you!

I did as directed and attempted to launch the tool challenge sim. I don't see any errors, but I'm also not seeing any PSM end effectors in the simulation.  This is what I see in the sim:

![ambf1](https://user-images.githubusercontent.com/56661595/236919890-8c27af69-7cda-4bb0-b406-0117b999bec1.png)
![ambf2](https://user-images.githubusercontent.com/56661595/236919894-2183b9a1-3e83-4b64-806b-6039935824f0.png)

**Comment by adnanmunawar** [comment]
Great, yes this is normal. The PSMs are retracted all the way back. You can run the scripts in the examples folder to control the arms.

**Reply by sunshineinsandiego** [reply]
Fantastic, thank you!  Also, is there a way to get the camera matrices for the L/R stereo cameras on the endoscope and subscribe to the L/R images separately from each of the cameras in this simulator?

**Reply by adnanmunawar** [reply]
By the camera matrices, do you mean the camera intrinsics or the projection matrix? You can construct either by taking into account the information from the camera specifications here (https://github.com/collaborative-robotics/surgical_robotics_challenge/blob/master/ADF/world/world_stereo.yaml#L42). You ll need the `near` and `far` plane, the `field view angle` and the `aspect ratio` which you can calculate from the `publish image resolution` values.

As far as we subscribing to the images from each camera, yes, each camera publishes to a separate image message.

**Reply by sunshineinsandiego** [reply]
Got it! Yes, was looking for the endoscope stereo L/R camera intrinsic matrices. In that YAML file I see Camera L, Camera L2, Camera R, and normal_camera. Which of these are the endoscope L/R cameras?

Thanks!

**Reply by adnanmunawar** [reply]
One can have many cameras (or any AMBF object) defined in the body of the ADF file but only the cameras (or objects) also defined at the top in the header list are the ones that are loaded (https://github.com/collaborative-robotics/surgical_robotics_challenge/blob/master/ADF/world/world_stereo.yaml#L3). With that being said, you want the output from `cameraL` and `cameraR`.

**Reply by sunshineinsandiego** [reply]
Terrific, thank you!
